flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/sqlserver/SqlServerDialect.java

```diff
@@ -0,0 +1,162 @@
+    public Optional<String> getUpsertStatement(
+            String tableName, String[] fieldNames, String[] uniqueKeyFields) {
+        List<String> nonUniqueKeyFields =
+                Arrays.stream(fieldNames)
+                        .filter(f -> !Arrays.asList(uniqueKeyFields).contains(f))
+                        .collect(Collectors.toList());
+        String fieldsProjection =
+                Arrays.stream(fieldNames)
+                        .map(this::quoteIdentifier)
+                        .collect(Collectors.joining(", "));
+
+        String valuesBinding =
+                Arrays.stream(fieldNames)
+                        .map(f -> ":" + f + " " + quoteIdentifier(f))
+                        .collect(Collectors.joining(", "));
+
+        String usingClause = String.format("SELECT %s", valuesBinding);
+        String onConditions =
+                Arrays.stream(uniqueKeyFields)
+                        .map(
+                                f ->
+                                        "[TARGET]."
+                                                + quoteIdentifier(f)
+                                                + "=[SOURCE]."
+                                                + quoteIdentifier(f))
+                        .collect(Collectors.joining(" AND "));
+        String updateSetClause =
+                nonUniqueKeyFields.stream()
+                        .map(
+                                f ->
+                                        "[TARGET]."
+                                                + quoteIdentifier(f)
+                                                + "=[SOURCE]."
+                                                + quoteIdentifier(f))
+                        .collect(Collectors.joining(", "));
+
+        String insertValues =
+                Arrays.stream(fieldNames)
+                        .map(f -> "[SOURCE]." + quoteIdentifier(f))
+                        .collect(Collectors.joining(", "));
+        return Optional.of(
+                String.format(
+                        "MERGE INTO %s AS [TARGET]"
+                                + " USING (%s) AS [SOURCE]"
+                                + " ON (%s)"
+                                + " WHEN MATCHED THEN"
+                                + " UPDATE SET %s"
+                                + " WHEN NOT MATCHED THEN"
+                                + " INSERT (%s) VALUES (%s);",
+                        quoteIdentifier(tableName),
+                        usingClause,
+                        onConditions,
+                        updateSetClause,
+                        fieldsProjection,
+                        insertValues));
+    }
+
+    @Override
+    public JdbcRowConverter getRowConverter(RowType rowType) {
+        return new SqlServerRowConverter(rowType);
+    }
+
+    @Override
+    public String getLimitClause(long limit) {
+        throw new IllegalArgumentException("SqlServerDialect does not support limit clause");
```

Comment:
Question to maintainer:

Is it ok to not support limit for SQL Server?
The reason is that SQL Server does not support limit natively, it supports by OFFSET syntax, but that syntax requires ORDER BY to be valid, with the current structure, I don't think we can support it properly.

This method has no contextual knowledge of whether ORDER BY exists in the larger statement or not.

If you think this is not acceptable, we might need to change code in `JdbcDynamicTableSource`, to support this special case

Reply:
IMO, +1 for ```not support in this PR now```.
I prefer to split this feature into a new PR. 
This should not be only related to dialect, but also need to change or introduce a new `pushdownXXX` in the relevant pushdown mechanism.
cc @bowenli86  @JingGe

Reply:
@qingwei91 Can we use `TOP` syntax ?

Reply:
Hi @maver1ck , I believe TOP syntax needs to be added like this:

SELECT TOP N FROM TABLE, this does not play well with current pattern where LIMIT is added to the end of the statement.

Maybe we should rethink this method api for it to be more generic? Is it ok to not include it in this PR? I am happy to take on this as a separate issue

Reply:
I agree. Probably we need to refactor getLimit part.
Let's make it separate PR.

Reply:
new ticket has been created: FLINK-29003